runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1867,6 +1873,28 @@ private void sendWorkerUpdatesToDataflowService(
           cumulativeCounters.extractUpdates(false, DataflowCounterUpdateExtractor.INSTANCE));
       counterUpdates.addAll(
           deltaCounters.extractModifiedDeltaUpdates(DataflowCounterUpdateExtractor.INSTANCE));
+      if (hasExperiment(options, "beam_fn_api")) {
+        while (!this.pendingMonitoringInfos.isEmpty()) {
+          final CounterUpdate item = this.pendingMonitoringInfos.poll();
+
+          // This change will treat counter as delta.
+          // This is required because we receive cumulative results from FnAPI harness,
+          // while streaming job is expected to receive delta updates to counters on same
+          // WorkItem.
+          if (item.getCumulative()) {
+            item.setCumulative(false);
+          } else {
+            // In current world all counters coming from FnAPI are cumulative.
+            // This is a safety check in case new counter type appears in FnAPI.
+            throw new UnsupportedOperationException(
```

Comment:
Shall we log the error and move on?

Reply:
It is better to fail fast in this case. New type of counter appears only when someone modifies SDK, so it comes from beam developer, not user.

Reply:
I think you should not fail in this case. What happens if a someone writes a custom SDK to run on Dataflow. It may introduce different metrics, but you should not fail.

An SDK can also be written by someone outside of beam repo entirely, not necessarily written by a beam developer.

https://docs.google.com/document/d/1MtBZYV7NAcfbwyy9Op8STeFNBxtljxgy69FkHMvhTMA/edit
Customization- Custom SDKs+engines can add their own custom metrics and custom monitored state without modifying beam proto files.

Reply:
That was bad way to formulate on my side:
CounterUpdates are generated within Java DF Worker. So if we receive different type of CounterUpdate here, that means that someone modified DF worker.